File: dp/数位dp/数位dp.js
/**
 * @param {string[]} digits
 * @param {number} n
 * @return {number}
 */
var atMostNGivenDigitSet = function(digits, n) {
    const set = new Set(digits);
    const nums = [];
    while (n > 0) {
        nums.unshift(n % 10);
        n = Math.floor(n / 10);
    }
    const m = nums.length;
    const memo = new Map();
    const dfs = (pre, idx, isLimit, isSkip) => {
        const key = [pre, idx, isLimit, isSkip].join(',');
        if (memo.has(key)) {
            return memo.get(key);
        }
        if (idx === m) {
            return isSkip ? 0 : 1;
        }
        let ret = 0;
        if (isSkip) {
            ret += dfs(pre, idx + 1, false, true);
        }
        const top = isLimit ? nums[idx] : 9;
        for (let i = isSkip ? 1 : 0; i <= top; i++) {
            if (!set.has(i.toString())) {
                continue;
            }
            ret += dfs(i, idx + 1, isLimit && i === top, false);
        }
        memo.set(key, ret);
        return ret;
    };

    // if `num1` is a string, get `+num1 - 1`
    // num1 = [...num1];
    // let i = num1.length - 1;
    // while (i >= 0 && num1[i] === '0') {
    //     i--;
    // }
    // num1[i] = String(+num1[i] - 1);
    // i++;
    // while (i < num1.length) {
    //     num1[i] = '9';
    //     i++;
    // }
    // num1 = num1.join('');

    const ans = dfs(0, 0, true, true);
    return ans;
};

File: dp/数位dp/数位dp.py
"""
数位dp模板，采用记忆化递归
902. 最大为 N 的数字组合
"""
from functools import cache
from typing import List
class Solution:
    def atMostNGivenDigitSet(self, digits: List[str], n: int) -> int:
        s = str(n)
        digits = set(digits)

        @cache
        def helper(i, is_num, is_limit):
            """ 数位DP模板
            Args:
                i: int 当前的位数
                is_num: bool 是否为数字，初始化为False
                is_limit: bool 上一位是否已达上界，初始化为True
            Returns:
                int
            """
            if i == len(s):
                return int(is_num)
            ret = 0
            if not is_num:
                ret = helper(i + 1, False, False)
            
            up = int(s[i]) if is_limit else 9
            for d in range(0 if is_num else 1, up + 1):
                if str(d) in digits:
                    ret += helper(i + 1, True, d == up and is_limit)
            return ret
        
        return helper(0, False, True)
